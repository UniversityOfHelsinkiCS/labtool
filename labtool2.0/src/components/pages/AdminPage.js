import React, { useEffect } from 'react'
import PropTypes from 'prop-types'
import { connect } from 'react-redux'
import { Button, Loader } from 'semantic-ui-react'
import { Redirect } from 'react-router'
import { getAllCI } from '../../services/courseInstance'
import { resetLoading } from '../../reducers/loadingReducer'
import { getAllUsers, updateOtherUser } from '../../services/user'
import { clearNotifications } from '../../reducers/notificationReducer'
import { usePersistedState } from '../../hooks/persistedState'
import { AdminCourseList } from './AdminPage/AdminCourseList'
import { AdminUserList } from './AdminPage/AdminUserList'
import DocumentTitle from '../DocumentTitle'
import Error from '../Error'

export const AdminPage = props => {
  const state = usePersistedState('AdminPage', {
    coursesList: false,
    usersList: false
  })

  useEffect(() => {
    props.clearNotifications()
  }, [])

  useEffect(() => {
    if (state.coursesList) {
      props.resetLoading()
      props.getAllCI()
    }
  }, [state.coursesList])

  useEffect(() => {
    if (state.usersList) {
      props.resetLoading()
      props.getAllUsers()
    }
  }, [state.usersList])

  const documentTitle = <DocumentTitle title={'Admin'} />

  if (!props.user.user.sysop) {
    return <Redirect to={`/labtool`} />
  } else if (props.errors && props.errors.length > 0) {
    return <Error errors={props.errors.map(error => `${error.response.data} (${error.response.status} ${error.response.statusText})`)} />
  } else if (props.loading.loading) {
    return (
      <>
        <Loader active />
        {documentTitle}
      </>
    )
  }

  return (
    <>
      {documentTitle}
      <div className="AdminPage">
        <h2>Welcome to the admin interface.</h2>
        <Button compact className={`ui button`} onClick={() => (state.coursesList = !state.coursesList)}>
          {state.coursesList ? 'Hide courses' : 'Manage courses'}
        </Button>
        <br />
        <br />
        {state.coursesList && <AdminCourseList courses={props.courseInstance} />}
        <br />
        <hr />
        <br />
        <Button compact className={`ui button`} onClick={() => (state.usersList = !state.usersList)}>
          {state.usersList ? 'Hide users' : 'Manage users'}
        </Button>
        <br />
        {state.usersList && <AdminUserList me={props.user} users={props.users} updateOtherUser={props.updateOtherUser} />}
      </div>
    </>
  )
}

const mapStateToProps = state => {
  return {
    user: state.user,
    courseInstance: state.courseInstance,
    users: state.users,
    loading: state.loading,
    errors: Object.values(state.loading.errors)
  }
}

const mapDispatchToProps = {
  getAllCI,
  resetLoading,
  getAllUsers,
  updateOtherUser,
  clearNotifications
}

AdminPage.propTypes = {
  user: PropTypes.object.isRequired,
  courseInstance: PropTypes.oneOfType([PropTypes.array, PropTypes.object]),
  users: PropTypes.array.isRequired,
  loading: PropTypes.object.isRequired,

  getAllCI: PropTypes.func.isRequired,
  resetLoading: PropTypes.func.isRequired,
  getAllUsers: PropTypes.func.isRequired,
  updateOtherUser: PropTypes.func.isRequired,
  clearNotifications: PropTypes.func.isRequired,

  errors: PropTypes.array
}

export default connect(mapStateToProps, mapDispatchToProps)(AdminPage)
