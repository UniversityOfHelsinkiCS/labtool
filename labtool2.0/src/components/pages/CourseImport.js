import React, { useEffect } from 'react'
import PropTypes from 'prop-types'
import { Button, Header, Table, Loader, Form } from 'semantic-ui-react'
import { connect } from 'react-redux'
import { Link, Redirect } from 'react-router-dom'
import { getIsAllowedToImport, getImportableCourses, importCourses } from '../../services/courseImport'
import { resetLoading, addRedirectHook } from '../../reducers/loadingReducer'
import { HorizontalScrollable } from '../HorizontalScrollable'
import { formatCourseName } from '../../util/format'

/**
 *  Show all the courses in a single list.
 */
export const CourseImport = props => {
  useEffect(() => {
    // run on component mount
    props.resetLoading()
    props.getIsAllowedToImport()
    props.getImportableCourses()
  }, [])

  const handleSubmit = async e => {
    e.preventDefault()
    const data = props.importable
    if (data) {
      const courses = data.filter(entry => e.target['import' + entry.hid] && e.target['import' + entry.hid].checked)
      props.addRedirectHook({
        hook: 'COURSE_IMPORT_DO_IMPORT_'
      })
      await props.importCourses({ courses })
    }
  }

  const renderCourse = instance => (
    <Table.Row key={instance.hid}>
      <Table.Cell>
        <Form.Checkbox name={'import' + instance.hid} />
      </Table.Cell>
      <Table.Cell>{instance.shorterId}</Table.Cell>
      <Table.Cell>{formatCourseName(instance.cname, instance.hid, instance.starts)}</Table.Cell>
      <Table.Cell>{instance.instructor}</Table.Cell>
      <Table.Cell>{instance.europeanStart}</Table.Cell>
      <Table.Cell>{instance.europeanEnd}</Table.Cell>
    </Table.Row>
  )

  return (
    <div className="CourseImport">
      <div className="mainContainer">
        <Header as="h2">Import courses</Header>
        {props.loading.loading || !Array.isArray(props.importable) || !props.canImport ? (
          <Loader active />
        ) : props.loading.redirect ? (
          <Redirect to="/labtool/courses" />
        ) : props.importable.length ? (
          <Form onSubmit={handleSubmit}>
            <HorizontalScrollable>
              <Table singleLine color="yellow" style={{ overflowX: 'visible' }}>
                <Table.Header>
                  <Table.Row>
                    <Table.HeaderCell colSpan="1" key={-1}>
                      Import?
                    </Table.HeaderCell>
                    <Table.HeaderCell colSpan="1">Course ID</Table.HeaderCell>
                    <Table.HeaderCell colSpan="1">Course name</Table.HeaderCell>
                    <Table.HeaderCell colSpan="1">Instructor (original)</Table.HeaderCell>
                    <Table.HeaderCell colSpan="1">Course start date</Table.HeaderCell>
                    <Table.HeaderCell colSpan="1">Course end date</Table.HeaderCell>
                  </Table.Row>
                </Table.Header>

                <Table.Body>{props.importable.map(renderCourse)}</Table.Body>
              </Table>
            </HorizontalScrollable>

            <br />
            <br />

            <Button type="submit" className="ui green button" content="Import" />

            <br />
            <br />

            <Link to="/labtool/courses">
              <Button className="ui button" type="cancel">
                Cancel
              </Button>
            </Link>
          </Form>
        ) : (
          <div>
            <h4>There are no courses in Kurki to import</h4>

            <Link to="/labtool/courses">
              <Button className="ui button" type="cancel">
                Back
              </Button>
            </Link>
          </div>
        )}
      </div>
    </div>
  )
}

const mapStateToProps = (state, ownProps) => {
  return {
    ...ownProps,
    loading: state.loading,
    canImport: state.courseImport.canImport,
    importable: state.courseImport.importable
  }
}

const mapDispatchToProps = {
  resetLoading,
  getIsAllowedToImport,
  getImportableCourses,
  importCourses,
  addRedirectHook
}

CourseImport.propTypes = {
  loading: PropTypes.object.isRequired,
  canImport: PropTypes.bool,
  importable: PropTypes.array,

  resetLoading: PropTypes.func.isRequired,
  getIsAllowedToImport: PropTypes.func.isRequired,
  getImportableCourses: PropTypes.func.isRequired,
  importCourses: PropTypes.func.isRequired,
  addRedirectHook: PropTypes.func.isRequired
}

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(CourseImport)
