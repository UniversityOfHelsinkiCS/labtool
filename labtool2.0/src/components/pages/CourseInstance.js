import React from 'react'

const CourseInstance = ({ instance, handleFieldChange }) => {
  return(
    <li> {instance.name} <button onClick={handleFieldChange} name="courseInstanceId" value={instance.id}>Register</button>
    </li>
  )
}

export default CourseInstance