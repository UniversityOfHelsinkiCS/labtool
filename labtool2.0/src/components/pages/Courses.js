import React, { Component } from 'react'
import { Button, List, Container, Header, Table, Label } from 'semantic-ui-react'
import { connect } from 'react-redux'
import { Link } from 'react-router-dom'

import { getAllCI } from '../../services/courseInstance'

/**
 *  Show all the courses in a single list.
 */
class Courses extends Component {
  componentWillMount() {
    this.props.getAllCI()
  }

  render() {
    return (
      <div>
        <Container>
          <Header as="h2">Courses</Header>
          <Table singleLine color="yellow">
            <Table.Header>
              <Table.Row>
                <Table.HeaderCell colSpan="1"> </Table.HeaderCell>
                <Table.HeaderCell colSpan="1">Course id</Table.HeaderCell>
                <Table.HeaderCell colSpan="1">Course name</Table.HeaderCell>
                <Table.HeaderCell colSpan="1">Course start date</Table.HeaderCell>
                <Table.HeaderCell colSpan="2"> </Table.HeaderCell>
              </Table.Row>
            </Table.Header>

            <Table.Body>
              {this.props.courseInstance.map(instance => (
                <Table.Row key={instance.id}>
                  <Table.Cell>
                    <div>
                      {instance.active === true ? (
                        <Label ribbon style={{ backgroundColor: '#21ba45' }}>
                          Active
                        </Label>
                      ) : (
                          ''
                        )}
                    </div>
                  </Table.Cell>
                  <Table.Cell>{instance.shorterId} </Table.Cell>
                  <Table.Cell>
                    <strong>
                      <a href={`/labtool/courses/${instance.ohid}`}>{instance.name}</a>
                    </strong>
                  </Table.Cell>

                  <Table.Cell> {instance.europeanStart} </Table.Cell>
                  <Table.Cell textAlign="center">
                    <Button circular size="tiny" icon={{ name: 'eye', size: 'large', color: 'blue' }} as={Link} to={`/labtool/courses/${instance.ohid}`} />
                  </Table.Cell>
                </Table.Row>
              ))}
            </Table.Body>
          </Table>

          <div className="Instructions">
            <List>
              <List.Item icon={{ name: 'eye', color: 'blue' }} content="Show course page" />
              <List.Item icon={{ name: 'square', color: 'green' }} content="Course is activated" />
            </List>
          </div>
        </Container>
      </div>
    )
  }
}

const mapStateToProps = (state, ownProps) => {
  return {
    courseInstance: state.courseInstance
  }
}

export default connect(mapStateToProps, { getAllCI })(Courses)
