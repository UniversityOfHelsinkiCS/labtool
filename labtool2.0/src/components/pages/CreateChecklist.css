.deleteButton {
  float: right;
}

.addForm {
  margin: 0px 10px 10px 10px;
}

.numberField {
  width: 100px;
}

.textField {
  width: 100%;
}

.formField {
  margin-top: 10px;
}

.topOptions {
  margin-bottom: 15px;
}

.saveButtonText {
  font-size: 24px;
}

.deleteButtonText {
  font-size: 14px;
}