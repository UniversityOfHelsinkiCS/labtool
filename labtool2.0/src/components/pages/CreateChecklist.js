import React, { Component } from 'react'
import { connect } from 'react-redux'
import { Header, Input, Label, Button, Popup, Card, Dropdown, Loader } from 'semantic-ui-react'
import { showNotification } from '../../reducers/notificationReducer'
import { resetLoading } from '../../reducers/loadingReducer'
import { createChecklist, getOneChecklist } from '../../services/checklist'
import { getOneCI, getAllCI } from '../../services/courseInstance'
import { resetChecklist, changeField, addTopic, addRow, removeTopic, removeRow, castPointsToNumber } from '../../reducers/checklistReducer'
import './CreateChecklist.css'

export class CreateChecklist extends Component {
  constructor(props) {
    super(props)
    this.state = {
      week: undefined, // tracks value of week dropdown.
      copyCourse: undefined, // tracks value of course dropdown.
      topicName: '', // tracks value inputted into topic creation dialog box.
      rowName: '', // tracks value inputted into row creation dialog box.
      openAdd: '', // which addForm is currently open. '' denotes no open addForms. Only one addForm can be open at one time.
      courseDropdowns: [] // Dropdown options to show for copying checklist.
    }
  }

  componentWillMount = async () => {
    await this.props.resetLoading()
    this.props.resetChecklist()
    this.props.getAllCI()
  }

  componentDidUpdate(prevProps, prevState) {
    if (this.state.week !== prevState.week) {
      this.setState({
        courseDropdowns: this.createCourseDropdowns()
      })
    }
  }

  // Make api call to save checklist to database.
  handleSubmit = async e => {
    e.preventDefault()
    const weeks = this.props.selectedInstance.weekAmount
    const checklists = this.props.selectedInstance.finalReview ? weeks + 1 : weeks
    if (this.state.week <= 0 || this.state.week > checklists) {
      this.props.showNotification({
        message: 'Invalid week.',
        error: true
      })
      return
    }
    try {
      this.props.createChecklist({
        courseInstanceId: this.props.selectedInstance.id,
        week: this.state.week,
        checklist: this.props.checklist.data
      })
    } catch (e) {
      this.props.showNotification({
        message: 'Could not parse JSON.',
        error: true
      })
    }
  }

  changeWeek = async (e, { value }) => {
    const week = value
    let copyCourse
    if (this.state.copyCourse) {
      if (week > this.props.selectedInstance.weekAmount) {
        copyCourse = this.props.courses.find(course => course.id === this.state.copyCourse).finalReview ? this.state.copyCourse : undefined
      } else {
        copyCourse = this.props.courses.find(course => course.id === this.state.copyCourse).weekAmount >= week ? this.state.copyCourse : undefined
      }
    } else {
      copyCourse = undefined
    }
    this.setState({
      week,
      copyCourse
    })
    this.loadChecklist(week)
  }

  changeCopyCourse = async (e, { value }) => {
    const copyCourse = value
    this.setState({
      copyCourse
    })
  }

  changeField = (key, name, field) => async e => {
    this.props.changeField({
      key,
      name,
      field,
      value: e.target.value
    })
  }

  // Make api call to receive checklist from database.
  loadChecklist = async week => {
    this.props.getOneChecklist({
      week,
      courseInstanceId: this.props.selectedInstance.id
    })
  }

  copyChecklist = async e => {
    if (!this.state.copyCourse) return
    const week = this.state.week > this.props.selectedInstance.weekAmount ? this.props.courses.find(course => course.id === this.state.copyCourse).weekAmount + 1 : this.state.week
    this.props.getOneChecklist({
      week,
      courseInstanceId: this.state.copyCourse,
      copying: true
    })
  }

  /**
   * First call opens a dialog box.
   * Subsequent call actually adds the new topic.
   */
  newTopic = async e => {
    e.preventDefault()
    if (this.state.openAdd !== 'newTopic') {
      this.setState({
        openAdd: 'newTopic'
      })
      return
    }
    if (this.state.topicName === '') {
      this.props.showNotification({
        message: 'Topic name cannot be blank.',
        error: true
      })
      return
    }
    this.props.addTopic({
      key: this.state.topicName
    })
    this.setState({
      topicName: '',
      openAdd: ''
    })
  }

  /**
   * First call opens a dialog box.
   * Subsequent call actually adds the new row.
   */
  newRow = key => async e => {
    e.preventDefault()
    if (this.state.openAdd !== key) {
      this.setState({
        openAdd: key
      })
      return
    }
    if (this.state.rowName === '') {
      this.props.showNotification({
        message: 'Checkbox name cannot be blank.',
        error: true
      })
      return
    }
    this.props.addRow({
      key,
      name: this.state.rowName
    })
    this.setState({
      rowName: '',
      openAdd: ''
    })
  }

  changeTopicName = async e => {
    this.setState({
      topicName: e.target.value
    })
  }

  changeRowName = async e => {
    this.setState({
      rowName: e.target.value
    })
  }

  removeTopic = key => async e => {
    this.props.removeTopic({
      key
    })
  }

  removeRow = (key, name) => async e => {
    this.props.removeRow({
      key,
      name
    })
  }

  cancelAdd = async e => {
    this.setState({
      openAdd: ''
    })
  }

  castPointsToNumber = (key, name) => async e => {
    this.props.castPointsToNumber({
      key,
      name
    })
  }

  weekFilter = courses => {
    return courses.filter(course => this.state.week <= course.weekAmount)
  }
  finalFilter = courses => {
    return courses.filter(course => course.finalReview)
  }
  createCourseDropdowns = () => {
    if (!this.props.courses || !this.props.selectedInstance) return []
    const courses = this.state.week > this.props.selectedInstance.weekAmount ? this.finalFilter(this.props.courses) : this.weekFilter(this.props.courses)
    const options = courses.filter(course => this.props.selectedInstance.id !== course.id).map(course => {
      return {
        value: course.id,
        text: `${course.name} (${course.europeanStart})`
      }
    })
    return options
  }

  render() {
    return (
      <div className="CreateChecklist">
        <Header>{this.props.selectedInstance.name}</Header>
        <div className="editForm">
          <div className="topOptions">
            <Dropdown placeholder="Select Checklist" selection value={this.state.week} onChange={this.changeWeek} options={this.props.weekDropdowns} />
            <div className="copyForm">
              <Button type="button" onClick={this.copyChecklist} disabled={!this.state.copyCourse}>
                Copy
              </Button>
              <Dropdown
                className="courseDropdown"
                placeholder="Copy checklist from another course"
                selection
                value={this.state.copyCourse}
                onChange={this.changeCopyCourse}
                options={this.state.courseDropdowns}
              />
            </div>
          </div>
          {this.props.loading.loading ? (
            <Loader active />
          ) : this.state.week !== undefined ? (
            <div>
              <div>
                {Object.keys(this.props.checklist.data).map(key => (
                  <Card fluid color="red" key={key}>
                    <Card.Content>
                      <Header>
                        {key}{' '}
                        <Button className="deleteButton" type="button" color="red" size="tiny" onClick={this.removeTopic(key)}>
                          <div className="deleteButtonText">Delete topic</div>
                        </Button>
                      </Header>
                    </Card.Content>
                    {this.props.checklist.data[key].map(row => (
                      <Card.Content key={row.name}>
                        <Header>
                          {row.name}{' '}
                          <Button className="deleteButton" type="button" color="red" size="tiny" onClick={this.removeRow(key, row.name)}>
                            <div className="deleteButtonText">Delete checkbox</div>
                          </Button>
                        </Header>
                        <div className="formField">
                          <Label>Points</Label>
                          <Input
                            className="numberField"
                            type="number"
                            step="0.01"
                            value={row.points}
                            onChange={this.changeField(key, row.name, 'points')}
                            onBlur={this.castPointsToNumber(key, row.name)}
                          />
                        </div>
                        <div className="formField">
                          <Label>Text when checked</Label>
                          <Input className="textField" type="text" value={row.textWhenOn} onChange={this.changeField(key, row.name, 'textWhenOn')} />
                        </div>
                        <div className="formField">
                          <Label>Text when unchecked</Label>
                          <Input className="textField" type="text" value={row.textWhenOff} onChange={this.changeField(key, row.name, 'textWhenOff')} />
                        </div>
                      </Card.Content>
                    ))}
                    <form className="addForm" onSubmit={this.newRow(key)}>
                      {/*This, like all other addForms is here to funnel both the button press 
                        as well as a user pressing enter into the same function.*/}
                      <Popup trigger={<Button type="submit" circular icon={{ name: 'add' }} />} content="Add new checkbox" />
                      {this.state.openAdd === key ? (
                        <div>
                          <Label>Name</Label>
                          <Input className="newRowNameInput" type="text" value={this.state.rowName} onChange={this.changeRowName} />
                          <Button type="button" onClick={this.cancelAdd}>
                            Cancel
                          </Button>
                        </div>
                      ) : (
                        <div />
                      )}
                    </form>
                  </Card>
                ))}
                <form className="addForm" onSubmit={this.newTopic}>
                  <Popup trigger={<Button type="submit" circular icon={{ name: 'add', size: 'large' }} />} content="Add new topic" />
                  {this.state.openAdd === 'newTopic' ? (
                    <div>
                      <Label>Name</Label>
                      <Input className="newTopicNameInput" type="text" value={this.state.topicName} onChange={this.changeTopicName} />
                      <Button type="button" onClick={this.cancelAdd}>
                        Cancel
                      </Button>
                    </div>
                  ) : (
                    <div />
                  )}
                </form>
              </div>
              <form onSubmit={this.handleSubmit}>
                {/*This is a form with a single button instead of just a button because it doesn't work 
                  (doesn't call the function) as just a button with onClick.*/}
                <Button className="saveButton" type="submit" color="green" size="large">
                  <div className="saveButtonText">Save</div>
                </Button>
              </form>
            </div>
          ) : (
            <div />
          )}
        </div>
      </div>
    )
  }
}

const createWeekDropdowns = selectedInstance => {
  if (!selectedInstance) return []
  const options = []
  let week = 1
  while (week <= selectedInstance.weekAmount) {
    options.push({
      value: week,
      text: `Week ${week}`
    })
    week++
  }
  if (selectedInstance.finalReview) {
    options.push({
      value: week,
      text: 'Final Review'
    })
  }
  return options
}

const mapStateToProps = (state, ownProps) => {
  const selectedInstance = state.courseInstance.find(course => course.ohid === ownProps.courseId)
  return {
    selectedInstance: selectedInstance || {},
    weekDropdowns: createWeekDropdowns(selectedInstance),
    checklist: state.checklist,
    courses: state.courseInstance,
    loading: state.loading
  }
}

const mapDispatchToProps = {
  showNotification,
  resetLoading,
  createChecklist,
  getOneCI,
  getAllCI,
  getOneChecklist,
  resetChecklist,
  changeField,
  addTopic,
  addRow,
  removeTopic,
  removeRow,
  castPointsToNumber
}

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(CreateChecklist)
