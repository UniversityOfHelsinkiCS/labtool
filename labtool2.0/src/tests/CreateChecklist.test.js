import React from 'react'
import { CreateChecklist } from '../components/pages/CreateChecklist'
import { shallow } from 'enzyme'

describe('<CreateChecklist /> component', () => {
  let wrapper

  const courseInstance = [
    {
      id: 10011,
      name: 'Aineopintojen harjoitustyö: Tietorakenteet ja algoritmit',
      start: '2018-03-11T21:00:00.000Z',
      end: '2018-04-29T21:00:00.000Z',
      active: true,
      weekAmount: 7,
      weekMaxPoints: 3,
      currentWeek: 1,
      ohid: 'TKT20010.2018.K.A.1',
      createdAt: '2018-03-26T00:00:00.000Z',
      updatedAt: '2018-03-26T00:00:00.000Z',
      europeanStart: '11.03.2018',
      europeanEnd: '29.04.2018',
      shorterId: 'TKT20010'
    },
    {
      id: 10012,
      name: 'Ohjelmistotekniikan menetelmät',
      start: '2018-03-11T21:00:00.000Z',
      end: '2018-04-29T21:00:00.000Z',
      active: true,
      weekAmount: 7,
      weekMaxPoints: 3,
      currentWeek: 1,
      ohid: 'TKT20002.2018.K.K.1',
      createdAt: '2018-03-26T00:00:00.000Z',
      updatedAt: '2018-03-26T00:00:00.000Z',
      europeanStart: '11.03.2018',
      europeanEnd: '29.04.2018',
      shorterId: 'TKT20002'
    },
    {
      id: 10013,
      name: 'Aineopintojen harjoitustyö: Tietokantasovellus',
      start: '2018-01-16T21:00:00.000Z',
      end: '2018-03-10T21:00:00.000Z',
      active: false,
      weekAmount: 5,
      weekMaxPoints: 3,
      currentWeek: 1,
      ohid: 'TKT20011.2018.K.A.1',
      createdAt: '2018-03-26T00:00:00.000Z',
      updatedAt: '2018-03-26T00:00:00.000Z',
      europeanStart: '16.01.2018',
      europeanEnd: '10.03.2018',
      shorterId: 'TKT20011'
    }
  ]

  const checklist = {
    data: {
      Koodi: [
        {
          name: 'Koodin laatu',
          points: 2,
          textWhenOn: 'Koodin laatu kiitettävää',
          textWhenOff: 'Koodin laadussa parantamisen varaa'
        }
      ]
    }
  }

  const loading = {
    loading: false,
    loadingHooks: [],
    redirect: false,
    redirectHooks: [],
    redirectFailure: false
  }

  const weekChoice = 6

  let mockFn = jest.fn()

  beforeEach(() => {
    wrapper = shallow(
      <CreateChecklist
        courses={courseInstance}
        courseId={courseInstance[0].ohid}
        selectedInstance={courseInstance[0]}
        checklist={checklist}
        loading={loading}
        weekDropdowns={[]}
        showNotification={mockFn}
        resetLoading={mockFn}
        createChecklist={mockFn}
        getOneCI={mockFn}
        getOneChecklist={mockFn}
        resetChecklist={mockFn}
        changeField={mockFn}
        addTopic={mockFn}
        addRow={mockFn}
        removeTopic={mockFn}
        removeRow={mockFn}
        getAllCI={mockFn}
      />
    )
    wrapper.setState({
      week: weekChoice
    })
  })

  it('renders without error', () => {
    expect(wrapper.find('.CreateChecklist').exists()).toEqual(true)
  })

  it('should render correctly', () => {
    expect(wrapper).toMatchSnapshot()
  })

  describe('Editing form', () => {
    it('autofills text areas', () => {
      const textInputs = wrapper.find('.textField')
      Object.keys(checklist.data).forEach(key => {
        checklist.data[key].forEach(row => {
          let textWhenOnMatch = undefined
          let textWhenOffMatch = undefined
          textInputs.forEach(ti => {
            if (ti.prop('value') === row.textWhenOn) {
              textWhenOnMatch = ti
            }
            if (ti.prop('value') === row.textWhenOff) {
              textWhenOffMatch = ti
            }
          })
          expect(textWhenOnMatch).not.toBe(undefined)
          expect(textWhenOffMatch).not.toBe(undefined)
        })
      })
    })

    it('autofills point values', () => {
      const pointInputs = wrapper.find('.numberField')
      Object.keys(checklist.data).forEach(key => {
        checklist.data[key].forEach(row => {
          let pointsMatch = undefined
          pointInputs.forEach(ti => {
            if (Number(ti.prop('value')) === row.points) {
              pointsMatch = ti
            }
          })
          expect(pointsMatch).not.toBe(undefined)
        })
      })
    })

    describe('Copying form', () => {
      it('Renders a copy form', () => {
        expect(wrapper.find('.copyForm').exists()).toEqual(true)
      })

      it('Renders appropriate options for week', () => {
        const options = wrapper.find('.courseDropdown').prop('options')
        expect(options.length).toEqual(courseInstance.filter(course => course.weekAmount >= weekChoice && course !== courseInstance[0]).length)
      })
    })
  })

  // It turns out there's no way of testing form functionality in shallow rendering.
  // https://github.com/airbnb/enzyme/issues/308
})
