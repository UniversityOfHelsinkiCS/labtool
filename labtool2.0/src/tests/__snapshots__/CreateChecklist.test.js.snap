// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CreateChecklist /> component should render correctly 1`] = `
<div
  className="CreateChecklist"
>
  <Header>
    Aineopintojen harjoitustyö: Tietorakenteet ja algoritmit
  </Header>
  <div
    className="editForm"
  >
    <div
      className="topOptions"
    >
      <Dropdown
        additionLabel="Add "
        additionPosition="top"
        closeOnBlur={true}
        deburr={false}
        icon="dropdown"
        minCharacters={1}
        noResultsMessage="No results found."
        onChange={[Function]}
        openOnFocus={true}
        options={Array []}
        placeholder="Select Checklist"
        renderLabel={[Function]}
        searchInput="text"
        selectOnBlur={true}
        selectOnNavigation={true}
        selection={true}
        value={6}
        wrapSelection={true}
      />
      <div
        className="copyForm"
      >
        <Button
          as="button"
          disabled={true}
          onClick={[Function]}
          role="button"
          type="button"
        >
          Copy
        </Button>
        <Dropdown
          additionLabel="Add "
          additionPosition="top"
          className="courseDropdown"
          closeOnBlur={true}
          deburr={false}
          icon="dropdown"
          minCharacters={1}
          noResultsMessage="No results found."
          onChange={[Function]}
          openOnFocus={true}
          options={
            Array [
              Object {
                "text": "Ohjelmistotekniikan menetelmät (11.03.2018)",
                "value": 10012,
              },
            ]
          }
          placeholder="Copy checklist from another course"
          renderLabel={[Function]}
          searchInput="text"
          selectOnBlur={true}
          selectOnNavigation={true}
          selection={true}
          wrapSelection={true}
        />
      </div>
    </div>
    <div>
      <div>
        <Card
          color="red"
          fluid={true}
          key="Koodi"
        >
          <CardContent>
            <Header>
              Koodi
               
              <Button
                as="button"
                className="deleteButton"
                color="red"
                onClick={[Function]}
                role="button"
                size="tiny"
                type="button"
              >
                <div
                  className="deleteButtonText"
                >
                  Delete topic
                </div>
              </Button>
            </Header>
          </CardContent>
          <CardContent
            key="Koodin laatu"
          >
            <Header>
              Koodin laatu
               
              <Button
                as="button"
                className="deleteButton"
                color="red"
                onClick={[Function]}
                role="button"
                size="tiny"
                type="button"
              >
                <div
                  className="deleteButtonText"
                >
                  Delete checkbox
                </div>
              </Button>
            </Header>
            <div
              className="formField"
            >
              <Label>
                Points
              </Label>
              <Input
                className="numberField"
                onBlur={[Function]}
                onChange={[Function]}
                step="0.01"
                type="number"
                value={2}
              />
            </div>
            <div
              className="formField"
            >
              <Label>
                Text when checked
              </Label>
              <Input
                className="textField"
                onChange={[Function]}
                type="text"
                value="Koodin laatu kiitettävää"
              />
            </div>
            <div
              className="formField"
            >
              <Label>
                Text when unchecked
              </Label>
              <Input
                className="textField"
                onChange={[Function]}
                type="text"
                value="Koodin laadussa parantamisen varaa"
              />
            </div>
          </CardContent>
          <form
            className="addForm"
            onSubmit={[Function]}
          >
            <Popup
              content="Add new checkbox"
              keepInViewPort={true}
              on="hover"
              position="top left"
              trigger={
                <Button
                  as="button"
                  circular={true}
                  icon={
                    Object {
                      "name": "add",
                    }
                  }
                  role="button"
                  type="submit"
                />
              }
            />
            <div />
          </form>
        </Card>
        <form
          className="addForm"
          onSubmit={[Function]}
        >
          <Popup
            content="Add new topic"
            keepInViewPort={true}
            on="hover"
            position="top left"
            trigger={
              <Button
                as="button"
                circular={true}
                icon={
                  Object {
                    "name": "add",
                    "size": "large",
                  }
                }
                role="button"
                type="submit"
              />
            }
          />
          <div />
        </form>
      </div>
      <form
        onSubmit={[Function]}
      >
        <Button
          as="button"
          className="saveButton"
          color="green"
          role="button"
          size="large"
          type="submit"
        >
          <div
            className="saveButtonText"
          >
            Save
          </div>
        </Button>
      </form>
    </div>
  </div>
</div>
`;
